### PR TITLE
the game tictactoe finally

### PR DESCRIPTION
**Pull request by unidrelyn**
## Summary
The tic tac toe game is made, with props and states.

## Screenshot
image

##Card Trello
https://trello.com/c/KHeH2fCH/32-estructura-de-carpetas

### CODE REVIEW

File: React/game/src/app/components/App/index.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import React, { Component } from 'react';
+
+import Game from '../Screen/Game';
```
Comment (DLotero) [inline React/game/src/app/components/App/index.js]: We can use webpack's aliases for better import here! (Plus, they're used in our real projects so it's awesome to know about them) https://til.hashrocket.com/posts/38d2d28b6b-better-module-imports-with-webpack-aliases

File: React/game/src/app/components/App/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+class App extends Component {
+  render() {
+    return <Game />;
```
Comment (DLotero) [inline React/game/src/app/components/App/index.js]: this `App` can be a functional component

File: React/game/src/app/components/App/styles.scss [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+.app-intro {
+  font-size: large;
+}
+
+@keyframes app-logo-spin {
+  from { transform: rotate(0deg); }
+  to { transform: rotate(360deg); }
+}
```
Comment (DLotero) [inline React/game/src/app/components/App/styles.scss]: Are these styles used somewhere? I don't see the reference. If not, please delete the file!

File: React/game/src/app/components/Screen/Game/Board/index.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import React, { Component } from 'react';
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/Board/index.js]: Subcomponents (components inside other components) like `Game` and `Board` should follow the folder structure `ParentComponent/components/ChildrenComponent` so in this case, it should be `Game/components/Board`

File: React/game/src/app/components/Screen/Game/Board/index.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import Square from '../Square';
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/Board/index.js]: you can use your brand new knowledge on `aliases` here, for example :D

File: React/game/src/app/components/Screen/Game/Board/index.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+import styles from './styles.scss';
+
+class Board extends Component {
+  renderSquare = place => (
+    <Square position={place} value={this.props.board[place]} handler={this.props.mouse} />
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/Board/index.js]: you can use destructuring for `props` here!

File: React/game/src/app/components/Screen/Game/Square/index.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+class Square extends Component {
+  handleClick = () => {
+    this.props.handler(this.props.position);
+  }
+
+  render() {
+    return (
+      <button className={styles.square} onClick={this.handleClick}>
+        {this.props.value}
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/Square/index.js]: destructuring

File: React/game/src/app/components/Screen/Game/Square/index.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import React, { Component } from 'react';
+
+import styles from './styles.scss';
+
+class Square extends Component {
+  handleClick = () => {
+    this.props.handler(this.props.position);
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/Square/index.js]: destructuring

File: React/game/src/app/components/Screen/Game/Square/styles.scss [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  text-align: center;
+  width: 34px;
+}
+
+.square:focus {
+  outline: none;
+}
+
+.kbd-navigation .square:focus {
+  background: #DDD;
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/Square/styles.scss]: remember to put all the colors in a separate `colors` file

File: React/game/src/app/components/Screen/Game/index.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+      this.setState({
+        gameEnd: true,
+        turnWinner: result
+      });
+    }
+  };
+
+  winner = () => {
+    const serialWin = [
+      [0, 1, 2],
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/index.js]: you could extract this in a `constants.js` file in the `Game` folder

File: React/game/src/app/components/Screen/Game/index.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    history: [
+      {
+        board: Array(9).fill(''),
+        id: 0
+      }
+    ]
+  };
+
+  mouseClick = place => {
+    const history = this.state.history.slice(0, this.state.moves + 1);
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/index.js]: `destructuringgg`

File: React/game/src/app/components/Screen/Game/index.js [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    if (board[place] === '' && !this.state.gameEnd) {
+      board[place] = this.state.turn;
+      this.setState({
+        history: history.concat([
+          {
+            board,
+            id: history.length
+          }
+        ]),
+        turn: this.state.turn == 'X' ? 'O' : 'X',
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/index.js]: `X` and `O ` could be properties of the class

File: React/game/src/app/components/Screen/Game/index.js [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  render() {
+    let status;
+    if (this.state.gameEnd) {
+      status = `Winner: ${this.state.turnWinner}`;
+    } else {
+      status = `Next player: ${this.state.turn}`;
+    }
+
+    const history = this.state.history;
+    const current = history[this.state.moves];
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/index.js]: you could destructure `state` for this constants

File: React/game/src/app/components/Screen/Game/index.js [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+    const moves = history.map(step => {
+      const gameTurn = step.id ? `Go to game start ${step.id}` : `Go to move # $`;
```
Comment (DLotero) [inline React/game/src/app/components/Screen/Game/index.js]: you use `step.id` several times, you could create a constant for that (and destructuring :D)

**Review by DLotero — CHANGES_REQUESTED** [review]

File: React/game/src/app/components/App/index.js (hunk 2 above)
Comment (unidrelyn) [reply React/game/src/app/components/App/index.js]: 👍

Comment (wfolini) [inline React/game/src/app/components/App/index.js]: Same as @DLotero says. Check this https://github.com/Wolox/tech-guides/blob/master/react/style-guide.md#class-vs-reactcreateclass-vs-stateless

File: React/game/src/app/components/Screen/Game/Square/index.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
+import React, { Component } from 'react';
```
Comment (wfolini) [inline React/game/src/app/components/Screen/Game/Square/index.js]: Same about @DLotero says. Following this file structure rules -> https://github.com/Wolox/tech-guides/blob/master/react/style-guide.md#folder-structure Move this subcomponent to folder `.../screens/Game/components/Square/...`. Also, `screens` folder must be in minus

File: React/game/src/app/components/App/styles.scss (hunk 3 above)
Comment (unidrelyn) [reply React/game/src/app/components/App/styles.scss]: 👍 file was delete

File: React/game/src/app/components/Screen/Game/Square/styles.scss (hunk 9 above)
Comment (wfolini) [inline React/game/src/app/components/Screen/Game/Square/styles.scss]: +1. Move all colors constants to `src/scss/variables/_colors.scss`

File: React/game/src/app/components/Screen/Game/index.js [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    if (this.state.gameEnd) {
+      status = `Winner: ${this.state.turnWinner}`;
+    } else {
+      status = `Next player: ${this.state.turn}`;
+    }
+
+    const history = this.state.history;
+    const current = history[this.state.moves];
+
+    const moves = history.map(step => {
```
Comment (wfolini) [inline React/game/src/app/components/Screen/Game/index.js]: I'd move this `moves` as a arrow function of this class component, like `jumpTo` function, so we can avoid to declare a new function in each render call.

File: React/game/src/app/components/Screen/Game/styles.scss [hunk 17]
```diff
@@ -0,0 +1,18 @@
+body {
```
Comment (wfolini) [inline React/game/src/app/components/Screen/Game/styles.scss]: `body` is a global style, so u must move it to `scss/index.scss` file

File: React/game/src/app/components/Screen/Game/styles.scss [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+  font: 14px 'Century Gothic', Futura, sans-serif;
+  margin: 20px;
+}
+
+ol,
+ul {
```
Comment (wfolini) [inline React/game/src/app/components/Screen/Game/styles.scss]: also `ol` and `ul`

**Review by wfolini — CHANGES_REQUESTED** [review]

File: React/game/src/app/components/App/index.js (hunk 2 above)
Comment (unidrelyn) [reply React/game/src/app/components/App/index.js]: 👍

File: React/game/src/app/components/App/index.js (hunk 1 above)
Comment (unidrelyn) [reply React/game/src/app/components/App/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Board/index.js (hunk 5 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Board/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Square/index.js (hunk 15 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Square/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Board/index.js (hunk 5 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Board/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Square/styles.scss (hunk 9 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Square/styles.scss]: 👍

File: React/game/src/app/components/Screen/Game/index.js (hunk 10 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Board/index.js (hunk 6 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Board/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Square/index.js (hunk 7 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Square/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Square/index.js (hunk 8 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Square/index.js]: 👍

File: React/game/src/app/components/Screen/Game/index.js (hunk 11 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/index.js]: 👍

File: React/game/src/app/components/Screen/Game/index.js (hunk 12 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/index.js]: 👍

File: React/game/src/app/components/Screen/Game/index.js (hunk 13 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/index.js]: 👍

File: React/game/src/app/components/Screen/Game/index.js (hunk 14 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/index.js]: 👍

File: React/game/src/app/components/Screen/Game/Square/styles.scss (hunk 9 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/Square/styles.scss]: 👍

File: React/game/src/app/components/Screen/Game/index.js (hunk 16 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/index.js]: 👍

File: React/game/src/app/components/Screen/Game/styles.scss (hunk 17 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/styles.scss]: 👍

File: React/game/src/app/components/Screen/Game/styles.scss (hunk 18 above)
Comment (unidrelyn) [reply React/game/src/app/components/Screen/Game/styles.scss]: 👍

File: React/game/src/app/screen/Game/Board/index.js [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+import Square from '~square';
+
+import styles from './styles.scss';
+
+class Board extends Component {
+  renderSquare = place => {
+    const board = this.props.board;
+    const mouse = this.props.mouse;
+    return <Square position={place} value={board[place]} handler={mouse} />;
```
Comment (DLotero) [inline React/game/src/app/screen/Game/Board/index.js:10]: remeber to use `const { board, mouse} = this.props`

File: React/game/src/app/screen/Game/index.js [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+import Board from './Board';
+import styles from './styles.scss';
+import { serialWin } from './constants';
+
+class Game extends Component {
+  state = {
+    moves: 0,
+    gameEnd: false,
+    turn: 'X',
```
Comment (DLotero) [inline React/game/src/app/screen/Game/index.js]: you can use `this.playerOne` here

File: React/game/src/app/screen/Game/index.js [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+    history: [
+      {
+        board: Array(9).fill(''),
+        id: 0
+      }
+    ]
+  };
+
+  playerOne = 'X';
+  playerTwo = 'O';
```
Comment (DLotero) [inline React/game/src/app/screen/Game/index.js]: move these `playerOne` and `playerTwo` to the top of the class

File: React/game/src/app/screen/Game/index.js [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+            board,
+            id: history.length
+          }
+        ]),
+        turn: this.state.turn === this.playerOne ? this.playerTwo : this.playerOne,
+        moves: ++this.state.moves
+      });
+    }
+    const result = this.winner();
+    if (result === 'X') {
```
Comment (DLotero) [inline React/game/src/app/screen/Game/index.js]: use `this,playerOne` here

File: React/game/src/app/screen/Game/index.js [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+      this.setState({
+        gameEnd: true,
+        turnWinner: result
+      });
+    } else if (result === 'O') {
```
Comment (DLotero) [inline React/game/src/app/screen/Game/index.js]: user `this.playerTwo` here

File: React/game/src/app/screen/Game/index.js [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+
+  winner = () => {
+    const { history } = this.state;
+    const { moves } = this.state;
+    const board = history[moves - 1].board;
+    for (let i = 0; i < serialWin.length - 1; i++) {
+      if (
+        board[serialWin[i][0]] === board[serialWin[i][1]] &&
+        board[serialWin[i][1]] === board[serialWin[i][2]] &&
+        (board[serialWin[i][0]] === 'O' || board[serialWin[i][0]] === 'X')
```
Comment (DLotero) [inline React/game/src/app/screen/Game/index.js]: use `this.playerOne` and `this.playerTwo` here

**Review by DLotero — CHANGES_REQUESTED** [review]
Very nice!! some small changes and we are reeeeady to go!

File: React/game/src/app/screen/Game/Board/index.js (hunk 19 above)
Comment (unidrelyn) [reply React/game/src/app/screen/Game/Board/index.js:10]: 👍

File: React/game/src/app/screen/Game/index.js (hunk 20 above)
Comment (unidrelyn) [reply React/game/src/app/screen/Game/index.js]: 👍

File: React/game/src/app/screen/Game/index.js (hunk 21 above)
Comment (unidrelyn) [reply React/game/src/app/screen/Game/index.js]: 👍

File: React/game/src/app/screen/Game/index.js (hunk 22 above)
Comment (unidrelyn) [reply React/game/src/app/screen/Game/index.js]: 👍

File: React/game/src/app/screen/Game/index.js (hunk 23 above)
Comment (unidrelyn) [reply React/game/src/app/screen/Game/index.js]: 👍

File: React/game/src/app/screen/Game/index.js (hunk 24 above)
Comment (unidrelyn) [reply React/game/src/app/screen/Game/index.js]: 👍

**Review by DLotero — DISMISSED** [review]

File: React/game/src/app/components/App/styles.scss [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+  padding: 20px;
+}
+
+.app-title {
+  font-size: 1.5em;
+}
+
+.app-intro {
+  font-size: large;
+}
```
Comment (wfolini) [inline React/game/src/app/components/App/styles.scss]: This file must be deleted

File: React/game/src/app/screen/Game/index.js [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+
+    return (
+      <div className={styles.game}>
+        <div className={styles.gameboard}>
+          <Board mouse={this.mouseClick} board={current.board} />
+        </div>
+        <div className={styles.gameinfo}>
+          <div>{status}</div>
+          <ol>
+            {history.map(step => {
```
Comment (wfolini) [inline React/game/src/app/screen/Game/index.js]: Don't use arrow function inside `render` method. Check this why is a bad practice https://medium.freecodecamp.org/why-arrow-functions-and-bind-in-reacts-render-are-problematic-f1c08b060e36

**Review by wfolini — CHANGES_REQUESTED** [review]

File: React/game/src/app/components/App/styles.scss (hunk 25 above)
Comment (unidrelyn) [reply React/game/src/app/components/App/styles.scss]: 👍

File: React/game/src/app/screen/Game/index.js (hunk 26 above)
Comment (unidrelyn) [reply React/game/src/app/screen/Game/index.js]: 👍

File: React/game/src/app/screen/Game/index.js [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
```
Comment (wfolini) [inline React/game/src/app/screen/Game/index.js]: U need to extract this arrow function, the map arrow function

**Review by wfolini — CHANGES_REQUESTED** [review]

File: React/game/src/app/components/Screen/Game/Square/index.js (hunk 15 above)
Comment (wfolini) [reply React/game/src/app/components/Screen/Game/Square/index.js]: @unidrelyn Check again this. This change wasn't applied

**Review by wfolini — APPROVED** [review]

**Pull request merged**